### PR TITLE
TIMRLINKN-58 Create correct migration path in database

### PR DESCRIPTION
**Pull request by gugmaster**

### CODE REVIEW

File: timrlink.net.CLI/Program.cs [hunk 1]
```diff
@@ -131,33 +135,60 @@ private async Task ExportProjectTime(string connectionString, string from, strin
...
             {
-                Logger.LogInformation($"Running Database Migration... ({string.Join(", ", pendingMigrations)})");
-                await context.Database.MigrateAsync();
+                Logger.LogInformation("Running Database Migration... ({pendingMigrations})",
+                    string.Join(", ", pendingMigrations));
+                try
+                {
+                    await context.Database.MigrateAsync();
+                }
+                catch (SqlException exception)
```
Comment (daberni) [inline timrlink.net.CLI/Program.cs:166]: I would assume, that there also can be other `SqlException` which are not related to the assumed problem.

Can we concretize this a bit more?? And only do the following steps, when it's a clash of creating the concrete tables?

File: timrlink.net.CLI/Program.cs [hunk 2]
```diff
@@ -131,33 +135,60 @@ private async Task ExportProjectTime(string connectionString, string from, strin
...
+                {
+                    Logger.LogWarning($"SQLException occured so we consider the database was created without migration: {exception}");
+                    
+                    // When SQLException occurs we know that the database was already created before.
+                    // So we insert the first migration 20221020122606_InitialMigration manually
+                    var firstMigrationName = pendingMigrations.FirstOrDefault();
+                    var version = Assembly.GetAssembly(typeof(DbContext))?.GetName().Version;
```
Comment (daberni) [inline timrlink.net.CLI/Program.cs:173]: why should the Assembly be null?

File: timrlink.net.CLI/Program.cs [hunk 3]
```diff
@@ -131,33 +135,60 @@ private async Task ExportProjectTime(string connectionString, string from, strin
...
+                {
+                    await context.Database.MigrateAsync();
+                }
+                catch (SqlException exception)
+                {
+                    Logger.LogWarning($"SQLException occured so we consider the database was created without migration: {exception}");
+                    
+                    // When SQLException occurs we know that the database was already created before.
+                    // So we insert the first migration 20221020122606_InitialMigration manually
+                    var firstMigrationName = pendingMigrations.FirstOrDefault();
```
Comment (daberni) [inline timrlink.net.CLI/Program.cs:172]: no need for `OrDefault()` when the list must no be empty

File: timrlink.net.CLI/Program.cs (hunk 1 above)
Comment (daberni) [inline timrlink.net.CLI/Program.cs:166]: So you didn't find a proper way to check if tables exist already neither?

**Review by daberni — CHANGES_REQUESTED** [review]

File: timrlink.net.CLI/Program.cs (hunk 1 above)
Comment (gugmaster) [reply timrlink.net.CLI/Program.cs:166]: I didn't check because I think this is a smarter approach. Will check now.

Comment (daberni) [reply timrlink.net.CLI/Program.cs:166]: I tried to explain the issue with this comment: https://github.com/troii/timrlink.net/pull/37#discussion_r1066412977

File: timrlink.net.CLI/Program.cs [hunk 4]
```diff
@@ -154,40 +154,58 @@ private async Task ExportGroups(string connectionString)
...
-                    // So we insert the first migration 20221020122606_InitialMigration manually
-                    var firstMigrationName = pendingMigrations.FirstOrDefault();
-                    var version = Assembly.GetAssembly(typeof(DbContext))?.GetName().Version;
+                    var result = Convert.ToInt32(await command.ExecuteScalarAsync());
+                    tablesExist = result > 0;
+                }
+                
+                if (tablesExist && !migrationExists)
+                {
+                    Logger.LogInformation("Inserting first migration manually");
```
Comment (daberni) [inline timrlink.net.CLI/Program.cs:184]: I think logging can be a bit more verbose, something like:

```Database already has existing tables before running migrations, marking initial migration as already done```

File: timrlink.net.CLI/Program.cs [hunk 5]
```diff
@@ -154,40 +154,58 @@ private async Task ExportGroups(string connectionString)
...
+                    // migrations. So we insert the first migration 20221020122606_InitialMigration manually
+                    // Then we can switch to migrations managed by Entity Framework
+                    var firstMigrationName = pendingMigrations.First();
+                    var version = Assembly.GetAssembly(typeof(DbContext)).GetName().Version;
                     if (firstMigrationName != null && version != null)
                     {
                         var versionString = $"{version.Major}.{version.Minor}.{version.Build}";
-                        
+
+                        await context.Database.ExecuteSqlRawAsync($"CREATE TABLE {efMigrationsHistoryTable}(MigrationId nvarchar(150) NOT NULL CONSTRAINT PK___EFMigrationsHistory PRIMARY KEY, ProductVersion nvarchar(32) NOT NULL);");
```
Comment (daberni) [inline timrlink.net.CLI/Program.cs:195]: shouldn't this table be already created because of the call to `context.Database.GetPendingMigrationsAsync()`?

**Review by daberni — APPROVED** [review]

**Comment by gugmaster** [comment]
On thing left. When we start from scratch with new approach - migrations first - database has to exist. This was not the case with the old approach because database was also created if it did not exist. Not sure if there is a way we can approach this. Let's talk tomorrow.

**Pull request merged**